gtsfm/densify/mvs_patchmatchnet.py

```diff
@@ -311,4 +355,83 @@ def filter_depth(
 
         dense_points = np.concatenate(vertices, axis=0)
         dense_point_colors = np.concatenate(vertex_colors, axis=0)
-        return dense_points, dense_point_colors
+
+        # compute and collect metrics during filtering points from depth maps
+        filtering_metrics = []
+        # compute the proportion of valid pixels in the geometric masks among all reference views
+        filtering_metrics.append(GtsfmMetric(name="geometric_mask_valid_ratios", data=geo_mask_ratios))
+        # compute the proportion of valid pixels in the confidence masks among all reference views
+        filtering_metrics.append(GtsfmMetric(name="confidence_mask_valid_ratios", data=conf_mask_ratios))
+        # compute the proportion of valid pixels in the joint masks among all reference views
+        filtering_metrics.append(GtsfmMetric(name="joint_mask_valid_ratios", data=joint_mask_ratios))
+        filtering_metrics.append(
+            GtsfmMetric(name="reprojection_errors", data=reprojection_errors, store_full_data=False)
+        )
+
+        return dense_points, dense_point_colors, GtsfmMetricsGroup(name="filtering metrics", metrics=filtering_metrics)
+
+
+def compute_filtered_reprojection_error(
+    dataset: PatchmatchNetData,
+    ref_view: int,
+    src_views: List[int],
+    depth_list: Dict[int, np.ndarray],
+    max_reprojection_err: float,
+    joint_mask: np.ndarray,
+) -> List[float]:
+    """Compute reprojection errors of reference view pixels among all source views, filtered by joint mask
+    Detailed steps include:
+        1. Compute coordinates of each pixel in reference depth map at reference camera frame by ref_intrinsics.
+        2. Project these points to a source view by ref_extrinsics, src_extrinsics and src intrinsics.
+        3. Use the source depth map to calculate the depths of reprojected pixel by interpolation.
+        4. Compute coordinates of each reprojected pixel at source camera frame by src_intrinsics.
+        5. Reproject these points to the reference view by src_extrinsics, ref_extrinsics, and ref intrinsics.
+        6. Compute the reprojection errors by Euclidean distance and filter the errors by the joint mask.
+
+    Args:
+        dataset: an instance of PatchmatchData as the inference dataset
+        ref_view: reference view pm_i
+        src_views: list of source view pm_i
+        depth_list: list of 2D depth maps (H, W) from each reference view
```

Comment:
hi @codyly, i think this shape might be wrong in the docstring if we can do:
```python
    # get reference view estimated depth map
    ref_depth_est = depth_list[ref_view][0]
    # get the resolution of reference view depth map
    height, width = ref_depth_est.shape[:2]
```

Reply:
should it be N,H,W or C,H,W?

Reply:
should be 1,H,W

Reply:
Yes, (1, H, W) is correct. Thanks!